xarray/core/computation.py

```diff
@@ -343,8 +344,10 @@ def apply_dataset_ufunc(func, *args, **kwargs):
     list_of_coords = build_output_coords(args, signature, exclude_dims)
 
     args = [getattr(arg, 'data_vars', arg) for arg in args]
+
     result_vars = apply_dict_of_variables_ufunc(
-        func, *args, signature=signature, join=join, fill_value=fill_value)
+        func, *args, signature=signature, join=data_vars_join,
```

Comment:
You can make that work with `dataset_fill_value=0`:
```
In [11]: add(ds0, ds1, 'outer', 'outer', dataset_fill_value=0)
Out[11]:
<xarray.Dataset>
Dimensions:  (x: 3)
Coordinates:
  * x        (x) int64 0 1 2
Data variables:
    a        (x) float64 1.0 2.0 nan
    b        (x) float64 nan 99.0 3.0
```

Are you suggesting changing the default value to `0`? It's not obvious to me what is the best choice for the default `dataset_fill_value` is. Maybe we should remove the default value of `None` and raise a more descriptive error if this happens.